pkg/operator/ceohelpers/machine_api.go

```diff
@@ -0,0 +1,41 @@
+package ceohelpers
+
+import (
+	machinelistersv1beta1 "github.com/openshift/client-go/machine/listers/machine/v1beta1"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/utils/pointer"
+)
+
+// IsMachineAPIFunctionalWrapper returns a function for determining if the Machine API is functional.
+// As of today Machine API is functional when we find Machines resources in the Running state.
+//
+// Note:
+//  since this function needs to take into account only master machine objects
+//  make sure machineInformer contain only filtered data
+//  otherwise it might be expensive to react to every node update in larger installations
+func IsMachineAPIFunctionalWrapper(masterMachineInformer cache.SharedInformer, masterMachineLister machinelistersv1beta1.MachineLister) func() (bool, error) {
+	return func() (bool, error) {
+		if !masterMachineInformer.HasSynced() {
+			return false, nil
+		}
+		machines, err := masterMachineLister.List(labels.Everything())
+		if err != nil {
+			return false, err
+		}
+		if len(machines) == 0 {
+			return false, nil
+		}
+
+		// we expect all machines to be in the Running state
```

Comment:
This is incorrect, we don't expect _all_ machines to be Running, we expect at least one to be Running. There are many reasons why, during a functional Machine API, you might have a single Machine not running (eg you just added a new Machine, it might take minutes or even hours for it to become Running depending on the platform), where the rest are Running, so this would break the logic here.

Reading through the enhancement, I think this is easier to reason by looking at the non-functional API section:
> In each of the scenarios, either that are no Machines in the cluster or the Machines will be stuck in the
`Provisioning` phase. **In particular the absence of any `Running` Machine signals that the Machine API is non-functional.**

If we see at least 1 Running Machine, it is safe to assume that the Machine API is functional (the other Machines may just be validly still booting)

Reply:
thanks, will update the PR.

Reply:
The comment here needs updating still